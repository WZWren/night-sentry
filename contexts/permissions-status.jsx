export const LocalPermStatus = Object.freeze({
    INIT: 'init',
    GRANTED: 'granted',
    UNDETERMINED: 'undetermined',
    DENIED: 'denied',
});