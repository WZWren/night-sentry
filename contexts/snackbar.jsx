import { createContext, useContext, useState } from "react";
import { Snackbar } from "react-native-paper";

const SnackbarContext = createContext({});

export function useSnackbar() {
    return useContext(SnackbarContext);
}

/**
 * An abstraction of the Snackbar element used to display error/alert messages.
 * 
 * @param {*} children The Child components of SnackbarProvider
 * @returns The SnackbarContext.Provider encapsulating the setMessage hook.
 */
export function SnackbarProvider({ children }) {
    const [message, setMessage] = useState("");
    console.log("SnackbarProvider loading...");

    const handleDismiss = () => {
        setMessage("");
    }

    return (
        <SnackbarContext.Provider value={{ setMessage }}>
            { children }
            <Snackbar
                visible={message}
                onDismiss={handleDismiss}
                duration={10000}
                action={{
                    label: "Close",
                    onPress: handleDismiss
                }}>
                {message}
            </Snackbar>
        </SnackbarContext.Provider>
    );
}