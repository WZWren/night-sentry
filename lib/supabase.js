import 'react-native-url-polyfill/auto'
import { createClient } from "@supabase/supabase-js";
import AsyncStorage from '@react-native-async-storage/async-storage'

const projectUrl = process.env.PROJECT_URL
const projectKey = process.env.PROJECT_KEY

export const supabase = createClient(projectUrl, projectKey, {
    auth: {
        storage: AsyncStorage,
    }
});