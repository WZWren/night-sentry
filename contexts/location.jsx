import { useState, useEffect, createContext, useContext } from 'react';
import * as Location from 'expo-location';

import { LocalPermStatus } from './permissions-status';

const LocationContext = createContext({});

export function useLocation() {
    return useContext(LocationContext);
}

export function LocationProvider({children}) {
    const [permissionStatus, setPermissionStatus] = useState(null);
    const [location, setLocation] = useState(null);

    useEffect(() => {
        if (permissionStatus == LocalPermStatus.INIT) {
            (async () => {
                let { status } = await Location.requestForegroundPermissionsAsync();
                setPermissionStatus(status);
                if (status !== LocalPermStatus.GRANTED) {
                  console.log('Permission to access location was denied/not accepted yet.');
                  return;
                }
                let location = await Location.getCurrentPositionAsync({});
                setLocation(location);
                console.log(location.coords.latitude + ", " + location.coords.longitude);
            })();
        }
    }, [permissionStatus]);

    return (
        <LocationContext.Provider value={{ location, permissionStatus, setPermissionStatus }}>
            {children}
        </LocationContext.Provider>
    );
}