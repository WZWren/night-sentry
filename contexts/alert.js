/**
 * alert.js manages all the PubSub stuff for alerts and services.
 * Functions in this js file will be called back in auth during the login step.
 */

import { supabase } from "../lib/supabase";

export async function fetchPublishers(user, setDistress) {
    const { data, error } = await supabase.from("close_contacts")
        .select("publisher, name:publisher(first_name, last_name)")
        .eq("subscriber", user.id)
        .is("confirmed", true);

    if (error) {
        console.log(error.message);
        return;
    }

    // linear runtime is OK for this, as a user's close contacts should
    // on average be within the 10s range.
    for (let i = 0; i < data.length; i++) {
        // subscribe the app client to the channel.
        supabase.channel(data[i].publisher)
            .on('postgres_changes',
                { event: 'INSERT', schema: 'public', table: 'alerts', filter: `user_id=eq.${data[i].publisher}` },
                (payload) => {
                    setDistress(`Distress signal received from ${data[i].name.first_name} ${data[i].name.last_name}`);
                    console.log(payload);
                }
            ).subscribe();
        console.log("Subscribed to " + data[i].publisher)
    }
}

export async function unsubClient() {
    const { error } = await supabase.removeAllChannels();

    if (error) {
        console.log(error.message);
        return;
    }
    console.log("unsubscribed to all");
}