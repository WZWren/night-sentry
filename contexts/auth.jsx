import { useRouter, useSegments } from "expo-router";

import { createContext, useContext, useEffect, useState } from "react";
import { supabase } from "../lib/supabase";

const AuthContext = createContext({});

function useProtectedRoute(loggedIn) {
    const segments = useSegments();
    const router = useRouter();

    useEffect(() => {
        console.log("Entering protected route...");
        // this checks if the route is in the (auth) group of routes - in case
        // the user is somehow not on this route if the user is not logged in
        const inAuthGroup = segments[0] === "(auth)";
        if (!loggedIn && !inAuthGroup) {
            console.log("Routing to Login...");
            router.replace("/login");
        } else if (loggedIn && inAuthGroup) {
            console.log("Routing to Alert...");
            router.replace("/alert");
        }
    }, [loggedIn, segments, router]);
}

/**
 * Parent component that provides the user value as part of its context.
 * 
 * @param {*} children - The Child Components of the AuthProvider
 * @returns The AuthContext.Provider component, encapsulating the user value.
 */
export function AuthProvider({ children }) {
    const [loggedIn, setLoggedIn] = useState(null);
    console.log("AuthProvider loading...");

    useProtectedRoute(loggedIn);

    // this sets up the useEffect hooks to redirect users to the correct page.
    useEffect(() => {
        const { data } = supabase.auth.onAuthStateChange((event, session) => {
            console.log("User changed! " + event);
            if (event === "SIGNED_IN") {
                setLoggedIn(session.user);
            } else if (event === "SIGNED_OUT") {
                setLoggedIn(null);
            }
        })
        return () => data.subscription.unsubscribe();
    }, []);

    return (
        <AuthContext.Provider value={{ loggedIn }}>{ children }</AuthContext.Provider>
    );
}