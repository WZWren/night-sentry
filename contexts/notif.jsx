// TODO: Move Notifications to its separate context.

import { createContext, useContext, useEffect, useState, useRef } from "react";
import { Platform } from "react-native";
import * as Device from 'expo-device';
import * as Notifications from 'expo-notifications';

import { supabase } from "../lib/supabase";
import { useAuth } from "./auth";

const NotificationsContext = createContext({});

export function useNotif() {
    return useContext(NotificationsContext);
}

// we handle the notification signup in the layout slot.
// notifications taken directly from the expo-notifications example.
Notifications.setNotificationHandler({
    handleNotification: async () => ({
        shouldShowAlert: true,
        shouldPlaySound: true,
        shouldSetBadge: false,
    }),
});

async function registerForPushNotificationsAsync() {
    let token;

    if (Platform.OS === 'android') {
        await Notifications.setNotificationChannelAsync('default', {
            name: 'default',
            importance: Notifications.AndroidImportance.MAX,
            vibrationPattern: [0, 250, 250, 250],
            lightColor: '#FF231F7C',
        });
    }

    if (Device.isDevice) {
        const { status: existingStatus } = await Notifications.getPermissionsAsync();
        let finalStatus = existingStatus;
        if (existingStatus !== 'granted') {
            const { status } = await Notifications.requestPermissionsAsync();
            finalStatus = status;
        }
        if (finalStatus !== 'granted') {
            console.log('Failed to get push token for push notification!');
            return;
        }
        token = (await Notifications.getExpoPushTokenAsync()).data;
        console.log(token);
    } else {
        console.log('Must use physical device for Push Notifications');
    }

    return token;
}

export function NotificationsProvider({ children }) {
    const { loggedIn } = useAuth();
    // expoPushToken is to be pushed to the supabase user_info table.
    const [ expoPushToken, setExpoPushToken ] = useState("");
    // notification is the physical notification from the app
    const [ notification, setNotification ] = useState(null);
    const notificationListener = useRef();
    const responseListener = useRef();

    useEffect(() => {
        registerForPushNotificationsAsync().then(token => setExpoPushToken(token));
    
        notificationListener.current = Notifications.addNotificationReceivedListener(notification => {
            setNotification(notification);
            console.log("Notification received! " + notification.request.content.title);
        });
    
        responseListener.current = Notifications.addNotificationResponseReceivedListener(response => {
            // TODO: this is where router linking will take place.
            console.log(response);
        });
    
        return () => {
            Notifications.removeNotificationSubscription(notificationListener.current);
            Notifications.removeNotificationSubscription(responseListener.current);
        };
    }, []);

    useEffect(() => {
        if (loggedIn && expoPushToken) {
            console.log("Updating user token: " + expoPushToken);
            (async () => {
                const { error } = await supabase.from("user_info")
                    .update({ expo_token: expoPushToken })
                    .eq("id", loggedIn.id);
                if (error) {
                    console.log(error.message);
                }
            })();
        }
    }, [loggedIn, expoPushToken]);

    return (
        <NotificationsContext.Provider value={{ notification, setNotification, expoPushToken }}>
            {children}
        </NotificationsContext.Provider>
    );
}