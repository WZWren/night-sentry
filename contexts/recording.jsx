import { createContext, useContext, useEffect, useState } from "react";
import { Audio } from "expo-av";
import { decode } from 'base64-arraybuffer';
import * as Notifications from 'expo-notifications';
import * as FileSystem from "expo-file-system";

import { useAuth } from "./auth";
import { supabase } from "../lib/supabase";
import { LocalPermStatus } from "./permissions-status";

const RecorderContext = createContext({});

export function useRecorder() {
    return useContext(RecorderContext);
}

export function RecorderProvider({children}) {
    const { loggedIn } = useAuth();
    const [ recording, setRecording ] = useState(null);
    const [ permissionStatus, setPermissionStatus ] = useState(null);

    useEffect(() => {
        if (permissionStatus == LocalPermStatus.INIT) {
            (async () => {
                let { status } = await Audio.requestPermissionsAsync();
                setPermissionStatus(status);
                if (status !== LocalPermStatus.GRANTED) {
                  console.log('Permission to access microphone was denied/not accepted yet.');
                  return;
                }
            })();
        }
    }, [permissionStatus])

    async function startRecording() {
        try {
            console.log('Requesting permissions..');
            await Audio.requestPermissionsAsync();
            await Audio.setAudioModeAsync({
                allowsRecordingIOS: true,
                playsInSilentModeIOS: true,
            });
    
            console.log('Starting recording..');
            const { recording } = await Audio.Recording.createAsync(Audio.RecordingOptionsPresets.HIGH_QUALITY);
            await Notifications.scheduleNotificationAsync({
                content: {
                    title: "Recording in progress...",
                    subtitle: "Tap to go to alerts page.",
                    autoDismiss: true,
                },
                identifier: "recording",
                trigger: null
            });
            setRecording(recording);
            console.log('Recording started');
        } catch (err) {
            console.error('Failed to start recording', err);
        }
    }

    async function stopRecording() {
        console.log('Stopping recording..');
        setRecording(undefined);
        await recording.stopAndUnloadAsync();
        await Audio.setAudioModeAsync({
          allowsRecordingIOS: false,
        });
        Notifications.dismissNotificationAsync("recording");
        const uri = recording.getURI();
        console.log('Recording stopped and stored at', uri);
        const file = await FileSystem.readAsStringAsync(uri, { encoding: FileSystem.EncodingType.Base64 });
        const { data, error } = await supabase.storage.from("audio").upload(`${loggedIn.id}/${Date.now()}.m4a`, decode(file), {
            contentType: 'audio/mpeg'
        });
        if (error) {
            console.log("failed to upload. " + error.message);
        }
        console.log(data);
    }

    return (
        <RecorderContext.Provider value={{permissionStatus, setPermissionStatus, startRecording, stopRecording, recording}}>
            {children}
        </RecorderContext.Provider>
    );
}