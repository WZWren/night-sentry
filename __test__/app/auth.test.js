import { Text, View } from "react-native";
import { screen, render, act, fireEvent } from "@testing-library/react-native";
import { renderRouter } from "expo-router/src/testing-library";
import { AuthProvider, useAuth } from "../../contexts/auth";
import * as Snackbar from "../../contexts/snackbar";
import { supabase } from "../../lib/supabase";
import LoginPage from "../../app/(auth)/login";
import Register from "../../app/(auth)/register";

const URL = process.env.PROJECT_URL;

const TestAuthComponent = () => {
    const { loggedIn } = useAuth();
    return (
        <View>
            <Text testID="log_value">{loggedIn?.id}</Text>
            <Text>testhere</Text>
        </View>
    );
}

test("Render the Login Page", async () => {
    const renderPage = render(<LoginPage/>);
    const tree = renderPage.toJSON();
    const signInText = await screen.findByText("Log in...");
    const signInBtn = await screen.findByText("Login");
    expect(signInText).toBeTruthy();
    expect(signInBtn).toBeTruthy();
    expect(tree).toMatchSnapshot();
    renderPage.unmount();
});

it("should not make a call to the server if fields are blank", async () => {
    const snackbar = jest.spyOn(Snackbar, "useSnackbar").mockImplementation(() => ({
        setMessage: (input) => {}
    }));
    const spy = jest.spyOn(supabase.auth, "signInWithPassword");
    const renderPage = render(<LoginPage/>);
    const input = "placeholder";

    const signInBtn = await screen.findByText("Login");
    fireEvent.press(signInBtn);

    expect(spy).not.toHaveBeenCalled();
})

it("should make a call to the supabase server on login", async () => {
    const snackbar = jest.spyOn(Snackbar, "useSnackbar").mockImplementation(() => ({
        setMessage: (input) => {}
    }));
    const spy = jest.spyOn(supabase.auth, "signInWithPassword");
    const renderPage = render(<LoginPage/>);
    const input = "placeholder";

    const mailField = await screen.findByPlaceholderText("Email");
    const passField = await screen.findByPlaceholderText("Password");
    const signInBtn = await screen.findByText("Login");
    fireEvent.changeText(mailField, input);
    fireEvent.changeText(passField, input);
    fireEvent.press(signInBtn);

    expect(spy).toHaveBeenCalled();

    renderPage.unmount();
    jest.clearAllMocks();
});

describe("Signup Page", () => {
    beforeAll(async () => {
        const snackbar = jest.spyOn(Snackbar, "useSnackbar").mockImplementation(() => ({
            setMessage: (input) => {}
        }));
    });

    it("should go to signup page when signup link is pressed", async () => {
        const renderPage = renderRouter();
        const signUpLink = await screen.findByText("Sign up");
        fireEvent.press(signUpLink);

        expect(renderPage.getSegments()[1]).toBe("register");
    });

    it("should not call supabase with empty inputs", async () => {
        render(<Register />);
        const spy = jest.spyOn(supabase.auth, "signUp");
        const signUpBtn = await screen.findByText("Sign-up");
        fireEvent.press(signUpBtn);

        expect(spy).not.toBeCalled();
        spy.mockReset();
    });

    it("should not call supabase if passwords do not match", async () => {
        render(<Register />);
        const spy = jest.spyOn(supabase.auth, "signUp");
        const input = "placeholder";
        const differingInput = "holderplace";
        const mailField = await screen.findByPlaceholderText("Email");
        const firstName = await screen.findByPlaceholderText("First Name");
        const lastName  = await screen.findByPlaceholderText("Last Name");
        const passField = await screen.findByPlaceholderText("Password");
        const repwField = await screen.findByPlaceholderText("Re-enter Password");
        fireEvent.changeText(mailField, input);
        fireEvent.changeText(firstName, input);
        fireEvent.changeText(lastName , input);
        fireEvent.changeText(passField, input);
        fireEvent.changeText(repwField, differingInput);

        const signUpBtn = await screen.findByText("Sign-up");
        fireEvent.press(signUpBtn);

        expect(spy).not.toBeCalled();

        fireEvent.changeText(mailField, "");
        fireEvent.changeText(firstName, "");
        fireEvent.changeText(lastName , "");
        fireEvent.changeText(passField, "");
        fireEvent.changeText(repwField, "");
        spy.mockReset();
    });

    it("should call the supabase server on signup, if all client-side regex passes", async () => {
        render(<Register />);
        const spy = jest.spyOn(supabase.auth, "signUp");
        const input = "placeholder";
        const mailField = await screen.findByPlaceholderText("Email");
        const firstName = await screen.findByPlaceholderText("First Name");
        const lastName  = await screen.findByPlaceholderText("Last Name");
        const passField = await screen.findByPlaceholderText("Password");
        const repwField = await screen.findByPlaceholderText("Re-enter Password");
        fireEvent.changeText(mailField, input);
        fireEvent.changeText(firstName, input);
        fireEvent.changeText(lastName , input);
        fireEvent.changeText(passField, input);
        fireEvent.changeText(repwField, input);

        const signUpBtn = await screen.findByText("Sign-up");
        fireEvent.press(signUpBtn);

        expect(spy).toBeCalled();

        fireEvent.changeText(mailField, "");
        fireEvent.changeText(firstName, "");
        fireEvent.changeText(lastName , "");
        fireEvent.changeText(passField, "");
        fireEvent.changeText(repwField, "");
        spy.mockReset();
    })
});


describe("AuthProvider", () => {
    it('provides logged-in data to the context as expected', async () => {
        const spy = jest.spyOn(global, 'fetch').mockImplementation(
            async (input) => {
                return new Response(JSON.stringify(response));
            }
        )

        const { getByTestId } = render(
            <AuthProvider>
                <TestAuthComponent />
            </AuthProvider>
        );

        await act(async () => {
            const { data } = await supabase.auth.signInWithPassword({ email: process.env.TEST_MAIL, password: "password" });
            if (data) {
                console.log(data);
            }
        });

        const textElement = getByTestId("log_value");
        console.log(textElement.children[0]);
        expect(textElement.children[0]).toBeTruthy();
    }, 20000);
});

const response = {
    access_token:process.env.TEST_JWT,
    token_type:"bearer",
    expires_in:3600,
    refresh_token:"testrefresh",
    user:{
        id:"test_id_here",
        aud:"authenticated",
        role:"authenticated",
        email:process.env.TEST_MAIL,
        email_confirmed_at:"2023-05-26T15:00:39.404616Z",
        phone:"",
        confirmed_at:"2023-05-26T15:00:39.404616Z",
        last_sign_in_at:"2023-07-15T09:53:41.745610054Z",
        app_metadata:{
            provider:"email",
            providers:["email"]
        },
        user_metadata:{},
        identities:[{
            id:"test_id_here",
            user_id:"test_id_here",
            identity_data:{
                email:process.env.TEST_MAIL,
                sub:"test_id_here",
            },
            provider:"email",
            last_sign_in_at:"2023-05-26T15:00:39.403327Z",
            created_at:"2023-05-26T15:00:39.40336Z",
            updated_at:"2023-05-26T15:00:39.40336Z"
        }],
        created_at:"2023-05-26T15:00:39.401139Z",
        updated_at:"2023-07-15T09:53:41.753379Z"
    }
}