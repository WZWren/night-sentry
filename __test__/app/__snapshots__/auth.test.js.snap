// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the Login Page 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "gap": 4,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "writingDirection": "ltr",
        },
        [
          {
            "fontFamily": "System",
            "fontSize": 32,
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 40,
          },
          undefined,
        ],
      ]
    }
  >
    Log in...
  </Text>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(231, 224, 236, 1)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        {
          "width": "80%",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(28, 27, 31, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "scaleY": 0.5,
            },
          ],
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      style={
        [
          {
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          {
            "minHeight": 56,
          },
        ]
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        cursorColor="rgba(103, 80, 164, 1)"
        editable={true}
        maxFontSizeMultiplier={1.5}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Email"
        placeholderTextColor="rgba(73, 69, 79, 1)"
        selectionColor="rgba(103, 80, 164, 1)"
        style={
          [
            {
              "margin": 0,
            },
            {
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            {
              "height": 56,
            },
            {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "color": "rgba(73, 69, 79, 1)",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "letterSpacing": 0.15,
              "lineHeight": undefined,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            [
              {},
            ],
            undefined,
          ]
        }
        testID="text-input-flat"
        textContentType="emailAddress"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(231, 224, 236, 1)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        {
          "width": "80%",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(28, 27, 31, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "scaleY": 0.5,
            },
          ],
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      style={
        [
          {
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          {
            "minHeight": 56,
          },
        ]
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        cursorColor="rgba(103, 80, 164, 1)"
        editable={true}
        maxFontSizeMultiplier={1.5}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Password"
        placeholderTextColor="rgba(73, 69, 79, 1)"
        secureTextEntry={true}
        selectionColor="rgba(103, 80, 164, 1)"
        style={
          [
            {
              "margin": 0,
            },
            {
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            {
              "height": 56,
            },
            {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "color": "rgba(73, 69, 79, 1)",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "letterSpacing": 0.15,
              "lineHeight": undefined,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            [
              {},
            ],
            undefined,
          ]
        }
        testID="text-input-flat"
        textContentType="password"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "transparent",
        "borderRadius": 20,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
    testID="button-container-outer-layer"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "transparent",
          "borderColor": "transparent",
          "borderRadius": 20,
          "borderStyle": "solid",
          "borderWidth": 0,
          "flex": undefined,
          "minWidth": 64,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 20,
            },
          ]
        }
        testID="button"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  [
                    {
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    false,
                    {
                      "marginHorizontal": 12,
                    },
                    undefined,
                    false,
                    {
                      "color": "rgba(103, 80, 164, 1)",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    {
                      "fontSize": 16,
                    },
                  ],
                ],
              ]
            }
            testID="button-text"
          >
            Login
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "gap": 4,
      }
    }
  >
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "fontFamily": "System",
            "fontWeight": "400",
            "letterSpacing": 0,
          },
          {
            "writingDirection": "ltr",
          },
          {
            "fontSize": 16,
          },
        ]
      }
    >
      No account?
    </Text>
    <Text
      accessibilityRole="link"
      href="/register"
      onPress={[Function]}
      style={
        {
          "color": "#00004D",
          "fontSize": 16,
        }
      }
    >
      Sign up
    </Text>
  </View>
</View>
`;
